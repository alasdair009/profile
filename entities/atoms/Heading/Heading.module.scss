@use "../../../app/styles/colors";
@use "../../../app/styles/dimensions";
@use "../../../app/styles/lineClamp";
@use "../../../app/styles/typography";
@use "../../../app/styles/helpers/rem";
@use "sass:map";

@mixin append-element($element-name) {
  @at-root #{selector-append($element-name, &)} {
    @content;
  }
}

.root {
  border: none !important;
  color: var(--color) !important;
  font-weight: map.get(typography.$fontWeights, bold);
  margin: 0 auto rem.toRem(map.get(dimensions.$sizes, s24)) !important;
  max-width: rem.toRem(dimensions.$globalTextMaxWidth);
  padding-bottom: 0;
  position: relative;
  text-align: var(--align);
  text-wrap: balance;
  width: fit-content;

  &Clamp {
    @include lineClamp.lineClamp;
  }

  &.Shadow {
    text-shadow: 0 0 rem.toRem(map.get(dimensions.$sizes, s4))
      colors.$colorblackEvil;
  }

  &::after {
    background: colors.$colorGreenGrass;
    bottom: 0;
    content: "";
    display: none;
    height: rem.toRem(3);
    left: 0;
    position: absolute;
    transform: rotate(-2deg);
    width: 100%;
  }

  &.h1 {
    max-width: rem.toRem(map.get(dimensions.$sizes, s512));
    padding-bottom: rem.toRem(map.get(dimensions.$sizes, s8));

    &::after {
      display: block;
    }
  }

  &.h2 {
    max-width: rem.toRem(map.get(dimensions.$sizes, s512));

    &::after {
      display: block;
    }
  }

  @each $headerElement, $headerSize in typography.$headerSizes {
    &.#{$headerElement} {
      font-size: rem.toRem($headerSize);
    }
  }
}
