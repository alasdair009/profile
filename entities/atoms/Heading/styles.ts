import styled from "styled-components";
import {CSSProperties} from "react";
import {globalTextMaxWidth, sizes} from "../../design-tokens/dimensions";
import {colors} from "../../design-tokens/colors";
import {rem} from "polished";
import {headingSizes, HeadingTypes} from "../../design-tokens/typography";
import {Property} from "csstype";

export const Root = styled.h1<{$color: Property.Color, $level: HeadingTypes, $align: Property.TextAlign}>`
  color: ${({$color}) => $color};
  font-size: ${({$level}) => headingSizes[$level]};
  margin: 0 auto ${sizes.s24.rem};
  max-width: ${rem(globalTextMaxWidth)};
  padding-bottom: ${({$level}) => $level === "h1" ? sizes.s8.rem : 0};
  position: relative;
  text-align: ${({$align}) => $align};
  width: fit-content;
  
  &::after {
    background: ${colors.greenGrass};
    bottom: 0;
    content: "";
    display: ${({$level}) => $level === "h1" ? "block" : "none"};
    height: ${rem(3)};
    left: 0;
    position: absolute;
    transform: rotate(-2deg);
    width: 100%;
  }
`;