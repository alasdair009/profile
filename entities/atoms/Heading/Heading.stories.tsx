import { Heading } from "./Heading";
import { Meta, StoryObj } from "@storybook/react";

const meta: Meta<typeof Heading> = {
  component: Heading,
  argTypes: {
    children: {
      control: {
        type: "text",
      },
    },
  },
};
export default meta;

export const Default: StoryObj<typeof Heading> = {
  args: {
    children: "Heading",
  },
};
