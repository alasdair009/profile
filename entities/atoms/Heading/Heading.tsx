import { Root } from "./styles";
import { HTMLAttributes, ReactNode } from "react";
import { colors, HeadingTypes, TextAlignment } from "@/entities";
import { Property } from "csstype";

export type HeadingProps = {
  /**
   * Visual style of the Heading.
   */
  level?: HeadingTypes;
  /**
   * Heading element level.
   */
  as?: HeadingTypes;
  /**
   * Contents of the heading.
   */
  children: ReactNode;
  /**
   * Text alignment of the heading.
   */
  align?: TextAlignment;
  /**
   * Color of the text.
   */
  color?: Property.Color;
  /**
   * Apply a subtle shadow behind the text.
   */
  textShadow?: boolean;
  /**
   * Number of lines the text is restricted to.
   */
  lines?: number;
} & HTMLAttributes<HTMLHeadingElement>;

/**
 * Text displayed in a heading at a customisable level visually and semantically.
 */
export function Heading({
  level = "h1",
  as,
  align = "left",
  color = colors.whiteGhost,
  textShadow = false,
  lines,
  children,
  ...rest
}: HeadingProps) {
  const headingAs = as ? as : level;
  return (
    <Root
      $color={color}
      $level={level}
      as={headingAs}
      $align={align}
      $textShadow={textShadow}
      $lines={lines}
      data-testid={Heading.name}
      {...rest}
    >
      {children}
    </Root>
  );
}
