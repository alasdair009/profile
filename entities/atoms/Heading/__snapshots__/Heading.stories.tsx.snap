// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`atoms/Heading H1 play-test 1`] = `
<h1 data-testid="Heading"
    class="sc-hDcvty hInCPP"
>
  Heading 1
</h1>
`;

exports[`atoms/Heading H2 play-test 1`] = `
<h2 data-testid="Heading"
    class="sc-hDcvty jRnwhu"
>
  Heading 2
</h2>
`;

exports[`atoms/Heading H3 play-test 1`] = `
<h3 data-testid="Heading"
    class="sc-hDcvty jxTMaz"
>
  Heading 3
</h3>
`;

exports[`atoms/Heading H4 play-test 1`] = `
<h4 data-testid="Heading"
    class="sc-hDcvty jxTMaz"
>
  Heading 4
</h4>
`;

exports[`atoms/Heading H5 play-test 1`] = `
<h5 data-testid="Heading"
    class="sc-hDcvty jxTMaz"
>
  Heading 5
</h5>
`;

exports[`atoms/Heading H6 play-test 1`] = `
<h6 data-testid="Heading"
    class="sc-hDcvty jxTMaz"
>
  Heading 6
</h6>
`;
